ujit_codegen.c

```diff
@@ -159,6 +159,36 @@ ujit_side_exit(jitstate_t* jit, ctx_t* ctx)
     return code_ptr;
 }
 
+#if RUBY_DEBUG
+
+#define GEN_COUNTER_INC(cb, counter_name) _gen_counter_inc(cb, &(ujit_runtime_counters . counter_name))
+static void
+_gen_counter_inc(codeblock_t *cb, int64_t *counter)
+{
+    if (!rb_ujit_opts.gen_stats) return;
+     mov(cb, REG0, const_ptr_opnd(counter));
+     gen_lock_prefix(cb); // for ractors.
+     add(cb, mem_opnd(64, REG0, 0), imm_opnd(1));
+}
+
+#define COUNTED_EXIT(existing, counter_name) _counted_side_exit(existing, &(ujit_runtime_counters . counter_name))
+// Increment a counter then take an existing side exit.
+static uint8_t *
+_counted_side_exit(uint8_t *existing_side_exit, int64_t *counter)
+{
+    if (!rb_ujit_opts.gen_stats) return existing_side_exit;
+
+    uint8_t *start = cb_get_ptr(ocb, ocb->write_pos);
+    _gen_counter_inc(ocb, counter);
+    jmp_ptr(ocb, existing_side_exit);
+    return start;
+}
+
+#else
+#define GEN_COUNTER_INC(cb, counter_name) ((void)0)
+#define COUNTED_EXIT(existing, counter_name) existing
```

Comment:
Should probably name the argument `side_exit` rather than existing. It's a bit confusing.